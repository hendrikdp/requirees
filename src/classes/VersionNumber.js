import {constants} from "../require-global";

class VersionNumber{

    constructor(version){
        const versionType = typeof version;
        if(versionType === 'string'){
            this.parse(version);
        }else if(versionType === 'object' && version !== null){
            const {tolerance, minor, major, patch, build, rc, isdefault, str} = version;
            this.tolerance = tolerance;
            this.minor = minor;
            this.major = major;
            this.patch = patch;
            this.build = build;
            this.rc = rc;
            this._isdefault = isdefault;
            this._str = str;
        }
    }

    parse(str){
        this._str = str.replace('-default', '');
        if(this._str === '*') this.tolerance = '*';
        const versionFragments = str.match(constants.reVersionNumber);
        if(str.indexOf('default')>-1) this._isdefault = true;
        if(versionFragments && versionFragments.length === 8){
            //check if this is a default package
            const indexOfDefault = versionFragments.indexOf('-default');
            if(indexOfDefault > -1) versionFragments.splice(indexOfDefault,1);
            //fill out the rest of the version object
            if(typeof versionFragments[1] === 'string') this.tolerance = versionFragments[1];
            this.major = VersionNumber.toNumber(versionFragments[2]);
            this.minor = VersionNumber.toNumber(versionFragments[3]);
            if(typeof versionFragments[4] === 'string') this.patch = VersionNumber.toNumber(versionFragments[4].substr(1));
            if(typeof versionFragments[5] === 'string') this.build = VersionNumber.toNumber(versionFragments[5].substr(1));
            if(typeof versionFragments[6] === 'string') this.rc = VersionNumber.toNumber(versionFragments[6].substr(1));
        }
    }

    static toNumber(str){
        const castedNum = Number(str);
        return isNaN(castedNum) ? str : castedNum;
    }

    get default(){
        return this._isdefault || false;
    }

    get str(){
        if(this._str) {
            return this._str;
        }else if(typeof this.major === 'undefined'){
            return 'anonymous';
        }else{
            const strTolerance = this.tolerance || '';
            const strPatch = typeof this.patch !== 'undefined' ? `.${this.patch}` : '';
            const strBuild = typeof this.build !== 'undefined' ? `.${this.build}` : '';
            const strRc = typeof this.rc !== 'undefined' ? `-${this.rc}` : '';
            return `${strTolerance}${this.major}.${this.minor}${strPatch}${strBuild}${strRc}`;
        }
    }

}

export default VersionNumber;