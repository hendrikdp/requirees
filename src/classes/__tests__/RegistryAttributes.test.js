import Attributes from '../RegistryAttributes.js';
import VersionNumber from '../VersionNumber.js';

describe('Registry Attributes', ()=>{

    const jqMock = {
        "url": "https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js",
        "version": new VersionNumber('3.4.1'),
        "type": 'js',
        "dependencies": undefined,
        "factory": undefined
    };
    const jqMock342rc3 = {
        "name": "myjquery",
        "url": "https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js",
        "version": new VersionNumber('3.4.2-rc3'),
        "type": "js",
        "dependencies": undefined,
        "factory": undefined
    };
    const jqMock3425 = {
        "name": "myjquery",
        "url": "https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js",
        "version": new VersionNumber('3.4.2.5'),
        "type": "js",
        "dependencies": undefined,
        "factory": undefined
    };

    test('Check if 1 url gets parsed the the correct object', ()=>{
        const {files} = new Attributes(['https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js']);
        jqMock.name = 'cdnjscloudflarecomajaxlibsjquerycore';
        expect(files[0]).toEqual({...jqMock, type:'js'});
    });

    test('Check if name is picked up correctly', ()=>{
        const {files} = new Attributes([{'myjquery': 'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}]);
        jqMock.name = 'myjquery';
        expect(files[0]).toEqual(jqMock);
    });

    test('Check if multiple arguments get picked up', ()=>{
        const {files} = new Attributes([{'myjquery':'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}]);
        jqMock.name = 'myjquery';
        expect(files[0]).toEqual(jqMock);
    });

    test('Check if multiple arguments get picked up and version can be overridden', ()=>{
        const {files} = new Attributes([{'myjquery': {url: 'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js', version: '3.4.2-rc3'}}]);
        expect(files[0]).toEqual(jqMock342rc3);
    });

    test('Check if multiple arguments get picked up and version can be overridden (order should not matter)', ()=>{
        const {files} = new Attributes([{'myjquery': {version:'3.4.2-rc3', url:'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}}]);
        expect(files[0]).toEqual(jqMock342rc3);
    });

    test('Check if parsed versions do not get touched', ()=>{
        const {files} = new Attributes([{'myjquery':{version: {major:3, minor:4, patch:2, build:5}, url: 'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}}]);
        files[0].version._str = files[0].version.str;
        expect(files[0]).toMatchObject(jqMock3425);
    });

    test('Check if complex structures are properly interpreted', ()=>{
        const {files} = new Attributes([{
            'myjquery': {
                version: '3.4.2.5',
                url: 'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'
            }
        }]);
        expect(files[0]).toEqual(jqMock3425);
    });

    test('Check if version is picked up from the package name', ()=>{
        const {files} = new Attributes([{'myjquery@3.4.2.5':'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}]);
        expect(files[0]).toEqual(jqMock3425);
    });

    test('Check if the first clear version number gets priority', ()=>{
        const {files} = new Attributes([{'myjquery@3.4.2.5':'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}]);
        expect(files[0]).toEqual(jqMock3425);
    });

    test('Check if name is provided, take the url', ()=>{
        const {files} = new Attributes(['./myjquery.js@3.4.2.85']);
        expect(files[0]).toMatchObject({
            name: 'myjquery',
            type: 'js',
            url: './myjquery.js',
            version: new VersionNumber('3.4.2.85')
        });
    });

    test('Check if no url is provided, take the name', ()=>{
        const {files} = new Attributes(['myjquery.js@3.4.2.85']);
        expect(files[0]).toMatchObject({
            name: 'myjquery',
            type: 'js',
            url: 'myjquery.js',
            version: new VersionNumber('3.4.2.85')
        });
    });

    test('Check if null gets ignored', ()=>{
        const {files} = new Attributes([{'myjquery@3.4.2.5':'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/core.js'}]);
        expect(files[0]).toEqual(jqMock3425);
    });

    test('Check if nothing fails if there is no version defined', ()=>{
        const {files} = new Attributes(['https://cdnjs.cloudflare.com/ajax/libs/jquery/core.js']);
        expect(files[0]).toMatchObject({
            "url": "https://cdnjs.cloudflare.com/ajax/libs/jquery/core.js",
            "name": "cdnjscloudflarecomajaxlibsjquerycore",
            "type": "js"
        });
    });

    test('Check if tolerance is calculated', ()=>{
        const {files} = new Attributes(['myjquery.js@^3.4.2.85']);
        expect(files[0]).toMatchObject({
            name: 'myjquery',
            type: 'js',
            url: 'myjquery.js',
            version: new VersionNumber('^3.4.2.85')
        })
    });

    test('Check if incorrect versions are not interpreted', ()=>{
        const {files} = new Attributes(['myjquery.js@^yeeeezzz']);
        expect(files[0]).toMatchObject({
            name: 'myjquery',
            type: 'js',
            url: 'myjquery.js',
            version: new VersionNumber('^yeeeezzz')
        });
    });


    test('Check if only major.minor gets interpreted', ()=>{
        const {files} = new Attributes(['myjquery.js@~3.4']);
        expect(files[0]).toMatchObject({
            name: 'myjquery',
            type: 'js',
            url: 'myjquery.js',
            version: {
                tolerance: '~',
                major: 3,
                minor: 4,
                str: '~3.4'
            }
        });
    });

});