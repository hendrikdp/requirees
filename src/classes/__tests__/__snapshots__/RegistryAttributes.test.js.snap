// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry Attributes Check if incorrect versions are not interpreted 1`] = `
Object {
  "dependencies": undefined,
  "factory": undefined,
  "name": "myjqueryjs@^yeeeezzz",
  "preventReregistration": undefined,
  "type": undefined,
  "url": "myjquery.js@^yeeeezzz",
  "version": VersionNumber {},
}
`;
